src/Language/Mulang/JSCompiler.hs

```diff
@@ -0,0 +1,358 @@
+{-# LANGUAGE QuasiQuotes, OverloadedStrings, ViewPatterns, TypeSynonymInstances, FlexibleInstances #-}
+
+module Language.Mulang.JSCompiler (toJS) where
+
+import           Language.Mulang.Ast
+import           Data.Text(pack, unpack, intercalate, empty, toLower, Text)
+import           Data.Maybe
+import           NeatInterpolation
+
+-- | Compiles a Mulang AST to an executable JS source.
+-- | If the AST is not representable in JS (for example, if it contains Facts or other non-OOP
+-- | abstractions), the function will yield Nothing.
+-- | The generated code asumes it will be appended to another source string containing all primitive definitions
+-- | (some sort of JRE for the parsed languaje).
+-- | Notice that, since some semantics of the original code's language are lost during the parsing,
+-- | executing this code might not be equivalent to executing the original sources.
+-- | It might be reasonable to drop this whole implementation in the future and replace it with a
+-- | direct interpretation of the Ast.
```

Comment:
Does it worth the effort? Which benefits would it yield? 

If performance is a concern - It is **not** right now actually, or not that much - what about using http://asmjs.org/ as an intermediate approach? It is a lower level target JavaScript but still a source-to-source approach.

Reply:
Nah, it's not about performance.

I guess it all ends up being a matter of preferences, but my point is that js is going to be a nice(ish) container for (some) OO model, but not so much for functional or logic. In my experience it's not that much harder to interpret an AST than compile it and it lets you have direct control on what should happen instead of having to worry about how JS will behave. It is also much cleaner to read :P and does not introduce dependencies to node to run tests properly.

If you really want src to src I don't think it matters what is the target language. JS is simple enough to be a good choice.

In any case not my call to make, but thought I should mention it.